apiserver/test/integration/clientset_test.go

```diff
@@ -60,6 +60,119 @@ func TestGroupVersion(t *testing.T) {
 	}
 }
 
+// TestPolicyWatch checks that the WatchManager closes watch when a new Tier is added
+func TestPolicyWatch(t *testing.T) {
+	const name = "test-policywatch"
+	rootTestFunc := func() func(t *testing.T) {
+		return func(t *testing.T) {
+			client, shutdownServer := getFreshApiserverAndClient(t, func() runtime.Object {
+				return &v3.GlobalNetworkPolicy{}
+			})
+			defer shutdownServer()
+			if err := testPolicyWatch(client); err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	if !t.Run(name, rootTestFunc()) {
+		t.Errorf("test-policywatch test failed")
+	}
+}
+
+func testPolicyWatch(client calicoclient.Interface) error {
+	order := float64(100.0)
+	tier := &v3.Tier{
+		ObjectMeta: metav1.ObjectMeta{Name: "custom-tier"},
+		Spec: v3.TierSpec{
+			Order: &order,
+		},
+	}
+
+	globalNetworkPolicy := &v3.GlobalNetworkPolicy{ObjectMeta: metav1.ObjectMeta{Name: "global-net-pol-watch"}}
+
+	w, err := client.ProjectcalicoV3().GlobalNetworkPolicies().Watch(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error watching GlobalNetworkPolicies (%s)", err)
+	}
+
+	// We should be able to receive this add event in our watch
+	_, err = client.ProjectcalicoV3().GlobalNetworkPolicies().Create(context.Background(), globalNetworkPolicy, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating GlobalNetworkPolicy (%s)", err)
+	}
+	defer func() {
+		_ = client.ProjectcalicoV3().GlobalNetworkPolicies().Delete(context.Background(), globalNetworkPolicy.Name, metav1.DeleteOptions{})
+	}()
+
+	done := sync.WaitGroup{}
+	done.Add(1)
+	timeout := time.After(5 * time.Second)
+	var timeoutErr error
+	var event watch.Event
+
+	go func() {
+		defer done.Done()
+		for {
+			select {
+			case event = <-w.ResultChan():
+				return
+			case <-timeout:
+				timeoutErr = fmt.Errorf("timed out waiting for events")
+				return
+			}
+		}
+	}()
+
+	// Creating a new Tier should force the watch to be closed
+	_, err = client.ProjectcalicoV3().Tiers().Create(context.Background(), tier, metav1.CreateOptions{})
```

Comment:
Should we wait here for Policy creation event is received before creating a tier (`done.Wait()`)? Is it possible sittuation when Tier could be created before the policy event is processed by the watcher? If that happens, the WatchManager will close the watch, and the goroutine will hit a timeout instead of receiving the event?